### PR TITLE
FINERACT-2081: Loan account data additional fields for summary and de…

### PR DESCRIPTION
**Pull request by alberto-art3ch**
…linquency

## Description

For including new fields in the Loan account data API, we need to modify the GET Loan by ID / external-id API to return those additional fields. API : /fineract-provider/api/v1/loans/{loanId}

[FINERACT-2081](https://issues.apache.org/jira/browse/FINERACT-2081)

- Delinquency data
Screenshot 2024-07-03 at 12 25 53 a m

- Summary data
Screenshot 2024-07-03 at 12 33 49 a m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 1]
```diff
@@ -91,9 +91,14 @@ public class LoanSummaryData {
     private BigDecimal totalCreditBalanceRefundReversed;
     private BigDecimal totalRepaymentTransaction;
     private BigDecimal totalRepaymentTransactionReversed;
+    private BigDecimal totalInterestRefund;
+    private BigDecimal totalInterestPaymentWaiver;
     private final Long chargeOffReasonId;
     private final String chargeOffReason;
 
+    private BigDecimal totalUnpaidAccruedDueInterest;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java:102]: Fields were created, but i dont see where these are calculated and sent back in the response!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java [hunk 2]
```diff
@@ -63,8 +63,7 @@ public enum LoanTransactionType {
     DOWN_PAYMENT(28, "loanTransactionType.downPayment"), //
     REAGE(29, "loanTransactionType.reAge"), //
     REAMORTIZE(30, "loanTransactionType.reAmortize"), //
-    INTEREST_PAYMENT_WAIVER(31, "loanTransactionType.interestPaymentWaiver"), //
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/LoanTransactionType.java]: unnecessary change... would you mind to revert?

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java [hunk 3]
```diff
@@ -58,7 +58,7 @@ private LocalDate getEffectiveDate() {
     }
 
     private boolean isBackdatedCharge() {
-        return loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate());
+        return (loanCharge.get().getDueDate() != null && loanCharge.get().getDueDate().isBefore(loanCharge.get().getSubmittedOnDate()));
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: Unnecessary change. Would you mind to revert? If it is a fix for a bug ticket, please extract it into a new PR!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 4]
```diff
@@ -616,6 +616,10 @@ private GetLoansLoanIdFeeFrequency() {}
             @Schema(example = "0.000000")
             public Double totalRepaymentTransaction;
             @Schema(example = "0.000000")
+            public Double totalInterestRefund;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: `totalUnpaidAccruedDueInterest` and `totalUnpaidAccruedNotDueInterest` are missing

File: fineract-provider/src/test/java/org/apache/fineract/portfolio/deliquency/DelinquencyWritePlatformServiceRangeChangeEventTest.java [hunk 5]
```diff
@@ -167,8 +167,9 @@ public void givenLoanAccountWithDelinquencyBucketWhenRangeChangeThenEventIsRaise
         LocalDate overDueSinceDate = DateUtils.getBusinessLocalDate().minusDays(2);
         LoanScheduleDelinquencyData loanScheduleDelinquencyData = new LoanScheduleDelinquencyData(1L, overDueSinceDate, 1L,
                 loanForProcessing);
-        CollectionData collectionData = new CollectionData(BigDecimal.ZERO, 2L, null, 2L, overDueSinceDate, BigDecimal.ZERO, null, null,
-                null, null, null, null);
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/portfolio/deliquency/DelinquencyWritePlatformServiceRangeChangeEventTest.java:171]: Would you mind to add unit or integration tests which validates whether the new fields are calculated and sent back with the right values? I dont see any "meaningful" tests :/

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 6]
```diff
@@ -4113,6 +4114,67 @@ public void uc137() {
         });
     }
 
+    // uc142: Loan Progressive processing type only with Advanced payment allocation
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Please fix the description of this use case

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 7]
```diff
@@ -4113,6 +4114,67 @@ public void uc137() {
...
+            Long repayment1TransactionId = addRepaymentForLoan(loanId, 251.0, "2 January 2023");
+            assertNotNull(repayment1TransactionId);
+            loanDetails = loanTransactionHelper.getLoanDetails(loanId);
+            LOG.info("Data {}", loanDetails.getSummary().getTotalUnpaidAccruedDueInterest());
+            LOG.info("Data {}", loanDetails.getSummary().getTotalUnpaidAccruedNotDueInterest());
+
+            Long repayment2TransactionId = addRepaymentForLoan(loanId, 251.0, "3 January 2023");
+            assertNotNull(repayment2TransactionId);
+            loanDetails = loanTransactionHelper.getLoanDetails(loanId);
+            LOG.info("Data {}", loanDetails.getSummary().getTotalUnpaidAccruedDueInterest());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Writing out with logger is not enough! Please user assertions to validate the proper amount was calculated and retrieved!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 8]
```diff
@@ -110,6 +116,11 @@ public static LoanSummaryData withTransactionAmountsSummary(final LoanSummaryDat
         BigDecimal totalCreditBalanceRefundReversed = BigDecimal.ZERO;
         BigDecimal totalRepaymentTransaction = BigDecimal.ZERO;
         BigDecimal totalRepaymentTransactionReversed = BigDecimal.ZERO;
+        BigDecimal totalInterestRefund = BigDecimal.ZERO;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: We dont need field yet (totalInterestRefund) -> hence not implemented yet

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 9]
```diff
@@ -191,4 +222,31 @@ private static BigDecimal computeTotalRepaymentTransactionAmount(Collection<Loan
                 loanTransactions);
         return totalRepaymentTransaction.add(totalDownPaymentTransaction);
     }
+
+    private static BigDecimal computeTotalAccruedDueAmount(Collection<LoanSchedulePeriodData> periods) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Please rename this to reflect it is calculating the accrued interest amount: `computeTotalInterestAccruedDueAmount`

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 10]
```diff
@@ -191,4 +222,31 @@ private static BigDecimal computeTotalRepaymentTransactionAmount(Collection<Loan
...
+                .map(period -> period.getTotalAccruedInterest()).reduce(BigDecimal.ZERO, BigDecimal::add);
+    }
+
+    private static BigDecimal computeTotalInterestPaidDueAmount(Collection<LoanSchedulePeriodData> periods) {
+        final LocalDate businessDate = DateUtils.getBusinessLocalDate();
+        return periods.stream().filter(period -> !period.getDownPaymentPeriod() && businessDate.isAfter(period.getDueDate()))
+                .map(period -> period.getInterestPaid()).reduce(BigDecimal.ZERO, BigDecimal::add);
+    }
+
+    private static BigDecimal fetchTotalAccruedNotDueAmount(Collection<LoanSchedulePeriodData> periods) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Please rename this to reflect it is calculating the accrued interest: `computeTotalInterestAccruedNotDueAmount`

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java [hunk 11]
```diff
@@ -191,4 +222,31 @@ private static BigDecimal computeTotalRepaymentTransactionAmount(Collection<Loan
...
+        final LocalDate businessDate = DateUtils.getBusinessLocalDate();
+        return periods.stream()
+                .filter(period -> !period.getDownPaymentPeriod() && period.getActualPeriod() && businessDate.isBefore(period.getDueDate()))
+                .map(period -> period.getTotalAccruedInterest()).reduce(BigDecimal.ZERO, BigDecimal::add);
+    }
+
+    private static BigDecimal fetchTotalInterestNotDueAmount(Collection<LoanSchedulePeriodData> periods) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Please rename this to reflect it is calculating the accrued interest: `computeTotalInterestNotDueAmount `

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanScheduleData.java [hunk 12]
```diff
@@ -54,6 +54,7 @@ public class LoanScheduleData {
     private final BigDecimal totalPaidLate;
     private final BigDecimal totalOutstanding;
     private final BigDecimal totalCredits;
+    private final BigDecimal totalAccruedInterest;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanScheduleData.java]: Do we really need this field?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanSchedulePeriodData.java [hunk 13]
```diff
@@ -72,7 +72,9 @@ public final class LoanSchedulePeriodData {
     private final BigDecimal totalActualCostOfLoanForPeriod;
     private final BigDecimal totalInstallmentAmountForPeriod;
     private final BigDecimal totalCredits;
+    private final BigDecimal totalAccruedInterest;
     private final Boolean downPaymentPeriod;
+    private final Boolean actualPeriod;
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanSchedulePeriodData.java]: Do we need this flag?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 14]
```diff
@@ -1160,8 +1160,8 @@ public String schema() {
...
                     + " ls.penalty_charges_amount as penaltyChargesDue, ls.penalty_charges_completed_derived as penaltyChargesPaid, ls.penalty_charges_waived_derived as penaltyChargesWaived, "
                     + " ls.penalty_charges_writtenoff_derived as penaltyChargesWrittenOff, ls.total_paid_in_advance_derived as totalPaidInAdvanceForPeriod, "
-                    + " ls.total_paid_late_derived as totalPaidLateForPeriod, ls.credits_amount as principalCredits, ls.credited_fee as feeCredits, ls.credited_penalty as penaltyCredits, ls.is_down_payment isDownPayment "
-                    + " from m_loan_repayment_schedule ls ";
+                    + " ls.total_paid_late_derived as totalPaidLateForPeriod, ls.credits_amount as principalCredits, ls.credited_fee as feeCredits, ls.credited_penalty as penaltyCredits, ls.is_down_payment isDownPayment, "
+                    + " ls.accrual_interest_derived as totalAccrualInterest " + " from m_loan_repayment_schedule ls ";
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Wouldnt it be better to name `InterestAccrued`?  (all the other interest related fields are `interestDue`, `interestPaid`, etc

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: Done! Added

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: Done! removed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanSchedulePeriodData.java (hunk 13 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanSchedulePeriodData.java]: Done! removed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 8 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! removed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 9 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! method renamed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 10 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! method renamed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 11 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Done! method renamed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanScheduleData.java (hunk 12 above)
Comment (alberto-art3ch) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/LoanScheduleData.java]: Done! attribute removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 14 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Done! renamed

File: fineract-provider/src/test/java/org/apache/fineract/portfolio/deliquency/DelinquencyWritePlatformServiceRangeChangeEventTest.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/test/java/org/apache/fineract/portfolio/deliquency/DelinquencyWritePlatformServiceRangeChangeEventTest.java:171]: ITC added, one for Delinquency data and other for accruals

**Review by alberto-art3ch — COMMENTED** [review]
Thanks!

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java (hunk 8 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanSummaryData.java]: Seems it was not removed, it is still there :(

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java (hunk 3 above)
Comment (adamsaghy) [reply fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/ChargeOrTransaction.java]: Nothing changed here either.... :/

**Review by adamsaghy — APPROVED** [review]
LGTM
